src/primitives/gf32.rs

```diff
@@ -0,0 +1,481 @@
+    /// Numeric value maps to bech32 character: 17 == "3".
+    pub const _3: Fe32 = Fe32(17);
+    /// Numeric value maps to bech32 character: 18 == "j".
+    pub const J: Fe32 = Fe32(18);
+    /// Numeric value maps to bech32 character: 19 == "n".
+    pub const N: Fe32 = Fe32(19);
+    /// Numeric value maps to bec32 character: 20 == "5".
+    pub const _5: Fe32 = Fe32(20);
+    /// Numeric value maps to bec32 character: 21 == "4".
+    pub const _4: Fe32 = Fe32(21);
+    /// Numeric value maps to bech32 character: 22 == "k".
+    pub const K: Fe32 = Fe32(22);
+    /// Numeric value maps to bech32 character: 23 == "h".
+    pub const H: Fe32 = Fe32(23);
+    /// Numeric value maps to bech32 character: 24 == "c".
+    pub const C: Fe32 = Fe32(24);
+    /// Numeric value maps to bech32 character: 25 == "e".
+    pub const E: Fe32 = Fe32(25);
+    /// Numeric value maps to bec32 character: 26 == "6".
+    pub const _6: Fe32 = Fe32(26);
+    /// Numeric value maps to bech32 character: 27 == "m".
+    pub const M: Fe32 = Fe32(27);
+    /// Numeric value maps to bech32 character: 28 == "u".
+    pub const U: Fe32 = Fe32(28);
+    /// Numeric value maps to bech32 character: 29 == "a".
+    pub const A: Fe32 = Fe32(29);
+    /// Numeric value maps to bec32 character: 30 == "7".
+    pub const _7: Fe32 = Fe32(30);
+    /// Numeric value maps to bech32 character: 31 == "l".
+    pub const L: Fe32 = Fe32(31);
+
+    /// Iterator over all field elements, in alphabetical order.
+    pub fn iter_alpha() -> impl Iterator<Item = Fe32> {
+        [
+            Fe32::A,
+            Fe32::C,
+            Fe32::D,
+            Fe32::E,
+            Fe32::F,
+            Fe32::G,
+            Fe32::H,
+            Fe32::J,
+            Fe32::K,
+            Fe32::L,
+            Fe32::M,
+            Fe32::N,
+            Fe32::P,
+            Fe32::Q,
+            Fe32::R,
+            Fe32::S,
+            Fe32::T,
+            Fe32::U,
+            Fe32::V,
+            Fe32::W,
+            Fe32::X,
+            Fe32::Y,
+            Fe32::Z,
+            Fe32::_0,
+            Fe32::_2,
+            Fe32::_3,
+            Fe32::_4,
+            Fe32::_5,
+            Fe32::_6,
+            Fe32::_7,
+            Fe32::_8,
+            Fe32::_9,
+        ]
+        .iter()
+        .copied()
+    }
+
+    /// Creates a field element from a single bech32 character.
+    pub fn from_char(c: char) -> Result<Fe32, Error> {
+        let byte = i8::try_from(u32::from(c)).map_err(|_| Error::InvalidChar(c))?;
+        let byte = byte as u8; // cast guaranteed to be ok since we started with an unsigned value
+        let u5 = u8::try_from(CHARS_INV[usize::from(byte)]).map_err(|_| Error::InvalidChar(c))?;
+        Ok(Fe32(u5))
+    }
+
+    /// Converts the field element to a lowercase bech32 character
+    pub fn to_char(self) -> char {
+        // casting and indexing fine as we have self.0 in [0, 32) as an invariant
+        CHARS_LOWER[self.0 as usize]
```

Comment:
In 310012dd3e5f5725b5bde9021606ed1d5542620d:

I think we can remove the cast and do `.into`. Since Rust 1.5 we've been able to convert from u8 this way (and since 1.26 from u16).